### PR TITLE
24 add new product zero coupon bond

### LINKED ISSUE #24: Add new product: Zero-coupon bond
**Context**
QuantSA does not have a zero-coupon bond product. We need this to model simple fixed-income instruments and to support basic pricing and discounting examples.

**Details**
Add a ZeroCouponBond class in QuantSA.Core.Prodcuts.Rates

### PR DESCRIPTION
**Pull request by rwx-oberholzer-marcelle**
# ✅ Pull Request Checklist

Before submitting your pull request, please ensure the following are complete:

## 🔍 Self Review
- [x] Confirmed that the destination branch is correct.
- [x] I have reviewed my own code carefully.
- [x] I have removed any debugging code or print statements.
- [x] Code changes are scoped to the purpose of this PR (no unrelated changes).
- [x] Comments are clear, and the code is self-explanatory where possible.
- [x] Documentation (if applicable) is updated.

## 📄 Task
- [x] The task/issue title is clear and descriptive.
- [x] The description explains **why** (context) and **what** (details) clearly.
- [x] Related issues are referenced (e.g., "Fixes #123").

## ✅ Code Quality
- [x] There are no new warnings or errors in the build output.
- [x] Code follows the project’s coding standards and formatting rules.

## 🧪 Testing
- [x] All unit and integration tests pass locally.
- [x] New functionality is covered by tests, or is explicitly marked as not testable.
- [x] Edge cases and error handling have been considered.

## ✅ CI
- [x] CI checks are passing (e.g., build, tests, linting).

## 📄 PR Quality
- [x] The PR title is clear and descriptive.
- [x] The PR description explains **what** and **why** clearly.
- [x] Related issues are referenced (e.g., "Fixes #123").

## 👥 Collaboration
- [x] I have assigned reviewers and/or requested feedback.
- [x] I’ve addressed or responded to all prior review comments (if applicable).

### CODE REVIEW

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+namespace QuantSA.Core.Tests.Products.Rates
+{
+    [TestClass]
+    public class ZeroCouponBondTests
+    {
+        [TestMethod]
+        public void ZeroCouponBond_ReturnsSingleCashflow()
+        {
+            var maturity = new Date(2028, 12, 31);
+            double notional = 1000000;
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: Use C# thousands separator syntax: 1000000 can be 1_000_000

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            var currency = new Currency("ZAR");
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: Check if there are any test utilities where you can just retrieve ZAR.

File: QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        public ZeroCouponBond(Date maturityDate, double notional, Currency currency)
+        {
+            _maturityDate = maturityDate;
+            _notional = notional;
+            _currency = currency;
+        }
+
+        public override List<Cashflow> GetCFs()
+        {
+            return new List<Cashflow>()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs]: () is not necessary.

File: QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            {
+                {
+                    new Cashflow(_maturityDate, _notional, _currency)
+                }
+            };
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs]: As per the ProductWrapper specification for GetCFs, filter with the valueDate and only return future cashflows.

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+using QuantSA.Core.Products.Rates;
+using QuantSA.Shared.Dates;
+using QuantSA.Shared.Primitives;
+
+namespace QuantSA.Core.Tests.Products.Rates
+{
+    [TestClass]
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs:9]: Add a test case for when it returns no cashflows.

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]

File: QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            if (_maturityDate > valueDate)
+            {
+                return new List<Cashflow>
+                {
+                    new Cashflow(_maturityDate, _notional, _currency)
+                };
+            }
+            return new List<Cashflow>();
+        }
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs]: This isn't on the interface of anything in the hierarchy, so defining it isn't very useful.

The trick here is that valueDate is defined on the parent and its access modifier is protected, so you have access to it in this class. You can move the filter you have here into GetCfs() and it should work.

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        public void ZeroCouponBond_NoCashflowsIfPastMaturity()
+        {
+            var maturity = new Date(2024, 01, 01);
+            double notional = 1_000_000;
+            var currency = TestHelpers.ZAR;
+
+            var bond = new ZeroCouponBond(maturity, notional, currency);
+
+            var valueDate = new Date(2024, 01, 02);  
+            var cashflows = bond.GetCFs(valueDate);
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: I don't recall the getCfs method taking in a value date, see my comment in your implementation.

What you need to do is set the valueDate on the product (its a member defined on the parent class and its protected so you have access to it), before you call GetCfs().

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+using QuantSA.Shared.Primitives;
+using QuantSA.Solution.Test;
+
+namespace QuantSA.Core.Tests.Products.Rates
+{
+    [TestClass]
+    public class ZeroCouponBondTests
+    {
+        [TestMethod]
+        public void ZeroCouponBond_ReturnsSingleCashflow()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: So the general pattern for the naming a test is **Function_[State]_ExpectedBehaviour**. I put State in square brackets because it may be optional. For this case then, a good name that meets this is **GetCfs_MaturityAfterValueDate_ReturnsExpectedCashflow**

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            var cashflows = bond.GetCFs();
+
+            Assert.AreEqual(1, cashflows.Count);
+            Assert.AreEqual(maturity, cashflows[0].Date);
+            Assert.AreEqual(notional, cashflows[0].Amount);
+            Assert.AreEqual(currency, cashflows[0].Currency);
+        }
+
+        [TestMethod]
+        public void ZeroCouponBond_NoCashflowsIfPastMaturity()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: A good name for this test is **GetCfs_MaturityBeforeValueDate_ReturnsNoCashflows**

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            var maturity = new Date(2024, 01, 01);
+            double notional = 1_000_000;
+            var currency = TestHelpers.ZAR;
+
+            var bond = new ZeroCouponBond(maturity, notional, currency);
+
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs:43]: Set the valueDate on the product before calling GetCfs()

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        [TestMethod]
+        public void ZeroCouponBond_ReturnsSingleCashflow()
+        {
+            var maturity = new Date(2028, 12, 31);
+            double notional = 1_000_000;
+            var currency = TestHelpers.ZAR;
+
+            var bond = new ZeroCouponBond(maturity, notional, currency);
+
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs:27]: Set the valueDate on the product before calling GetCfs()

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    [TestClass]
+    public class ZeroCouponBondTests
+    {
+        [TestMethod]
+        public void GetCfs_MaturityAfterValueDate_ReturnsExpectedCashflow()
+        {
+            var maturity = new Date(2028, 12, 31);
+            double notional = 1_000_000;
+            var currency = new Currency("ZAR");
+            //var currency = TestHelpers.ZAR;
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: You shouldn't leave commented code in the final PR.

Decide whether you're going to use the test utility or just specify the currency manually as you have in this setup.

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            Assert.AreEqual(currency, cashflows[0].Currency);
+        }
+
+        [TestMethod]
+        public void GetCfs_MaturityBeforeValueDate_ReturnsNoCashflows()
+        {
+            var maturity = new Date(2024, 01, 01);
+            double notional = 1_000_000;
+            var currency = new Currency("ZAR");
+            //var currency = TestHelpers.ZAR;
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: Don't leave test comments in the code.

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+{
+    [TestClass]
+    public class ZeroCouponBondTests
+    {
+        [TestMethod]
+        public void GetCfs_MaturityAfterValueDate_ReturnsExpectedCashflow()
+        {
+            var maturity = new Date(2028, 12, 31);
+            double notional = 1_000_000;
+            var currency = new Currency("ZAR");
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: Multiple test methods are using these values. In this case, declare them at the class level and set them up in the Setup method. This will allow you to reuse them across tests.

File: QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        public ZeroCouponBond(Date maturityDate, double notional, Currency currency)
+        {
+            _maturityDate = maturityDate;
+            _notional = notional;
+            _currency = currency;
+        }
+
+        public override void SetValueDate(Date valueDate)
+        {
+            this.valueDate = valueDate; // ✅ simple override for basic product
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core/Products/Rates/ZeroCouponBond.cs]: Necessary comment can be removed. Remember, comments are there to explain *why* you've done something, not what you've done. Writing readable code does the latter.

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]

File: QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs (hunk 14 above)
Comment (rwx-oberholzer-marcelle) [reply QuantSA/QuantSA.Core.Tests/Products/Rates/ZeroCouponBondTests.cs]: Thanks, I'll have to change this at more tests other than just this one.

**Review by rwx-jacquier-robert — APPROVED** [review]
